### ISSUE TITLE
Ubuntu 20.04/LXD 4.6 (snap) - unable to add serial/usb device to container "...is mounted nodev"

### ISSUE THREAD

**Issue by mhansi12** [lxc/lxd#8024]
# Required information

 * Distribution: Ubuntu
 * Distribution version: 20.04.1
 * The output of "lxc info" or if that fails:
   * Kernel version: 5.4.0-48-generic
   * LXC version: 4.0.4
   * LXD version: 4.6
   * Storage backend in use:  ZFS

# Issue description

Migrating containers from Ubuntu 16.05 / LXD 3.x to Ubuntu 20.04 / LXD 4.6

whenn adding a unix-char or usb device I get:

Error: Failed to start device "serial-port": Can't create device as devices path is mounted nodev

I modified the underlying ZFS to explicitly remove the "nodev" mount flag, but no change:
("zfs set devices=on rpool/ROOT/ubuntu_csxyom/var")

# cat /proc/mounts |grep var

rpool/ROOT/ubuntu_csxyom/var /var zfs rw,noatime,xattr,posixacl 0 0
rpool/ROOT/ubuntu_csxyom/var/cache /var/cache zfs rw,nosuid,nodev,noexec,noatime,xattr,posixacl 0 0
rpool/ROOT/ubuntu_csxyom/var/lib /var/lib zfs rw,noatime,xattr,posixacl 0 0
rpool/ROOT/ubuntu_csxyom/var/tmp /var/tmp zfs rw,nosuid,nodev,noatime,xattr,posixacl 0 0
rpool/ROOT/ubuntu_csxyom/var/log /var/log zfs rw,nosuid,nodev,noexec,noatime,xattr,posixacl 0 0
rpool/ROOT/ubuntu_csxyom/var/spool /var/spool zfs rw,nosuid,nodev,noexec,noatime,xattr,posixacl 0 0
rpool/ROOT/ubuntu_csxyom/var/lib/apt /var/lib/apt zfs rw,nosuid,nodev,noatime,xattr,posixacl 0 0
rpool/ROOT/ubuntu_csxyom/var/lib/dpkg /var/lib/dpkg zfs rw,nosuid,nodev,noatime,xattr,posixacl 0 0
tmpfs /var/snap/lxd/common/ns tmpfs rw,relatime,size=1024k,mode=700 0 0
nsfs /var/snap/lxd/common/ns/shmounts nsfs rw 0 0
nsfs /var/snap/lxd/common/ns/mntns nsfs rw 0 0

# Steps to reproduce

 1. Step one
  Install Ubuntu 20.04, Install LXD with snap, use ZFS als backend FS
 2. Step two
  create simple comainter : lxd launch ubuntu:focal tmpi
 3. Step three
  Add a usb ( or unix-char) device
   lxc config  device add tempi serial-port  usb  vendorid=0403 productid=6001

# Information to attach

 - [ ] Any relevant kernel output (`dmesg`)
 - [ ] Container log (`lxc info NAME --show-log`)
Name: meteo
Location: none
Remote: unix://
Architecture: i686
Created: 2020/10/11 17:39 UTC
Status: Running
Type: container
Profiles: default
Pid: 136614
Ips:
  eth0: inet    10.31.5.87      veth7205e6b6
  lo:   inet    127.0.0.1
Resources:
  Processes: 36
  Disk usage:
    root: 732.69MB
  CPU usage:
    CPU usage (in seconds): 30
  Memory usage:
    Memory (current): 113.39MB
    Memory (peak): 124.84MB
  Network usage:
    eth0:
      Bytes received: 5.85MB
      Bytes sent: 2.10MB
      Packets received: 18639
      Packets sent: 14144
    lo:
      Bytes received: 0B
      Bytes sent: 0B
      Packets received: 0
      Packets sent: 0

Log:

lxc meteo 20201012201906.381 WARN     cgfsng - cgroups/cgfsng.c:mkdir_eexist_on_last:1152 - File exists - Failed to create directory "/sys/fs/cgroup/cpuset//lxc.monitor.meteo"
lxc meteo 20201012201906.382 WARN     cgfsng - cgroups/cgfsng.c:mkdir_eexist_on_last:1152 - File exists - Failed to create directory "/sys/fs/cgroup/cpuset//lxc.payload.meteo"
lxc meteo 20201012201906.386 WARN     cgfsng - cgroups/cgfsng.c:fchowmodat:1573 - No such file or directory - Failed to fchownat(17, memory.oom.group, 1000000000, 0, AT_EMPTY_PATH | AT_SYMLINK_NOFOLLOW )

 - [ ] Container configuration (`lxc config show NAME --expanded`)
architecture: i686
config:
  boot.autostart: "true"
  boot.autostart.delay: "17"
  environment.TZ: Europe/Vienna
  image.architecture: i386
  image.description: ubuntu 16.04 LTS i386 (release) (20171221)
  image.label: release
  image.os: ubuntu
  image.release: xenial
  image.serial: "20171221"
  image.version: "16.04"
  volatile.base_image: e239ebf7aafd41075b6af107bcce224d8684542a42a8af184ffb596107c32aea
  volatile.idmap.base: "0"
  volatile.idmap.current: '[{"Isuid":true,"Isgid":false,"Hostid":1000000,"Nsid":0,"Maprange":1000000000},{"Isuid":false,"Isgid":true,"Hostid":1000000,"Nsid":0,"Maprange":1000000000}]'
  volatile.idmap.next: '[{"Isuid":true,"Isgid":false,"Hostid":1000000,"Nsid":0,"Maprange":1000000000},{"Isuid":false,"Isgid":true,"Hostid":1000000,"Nsid":0,"Maprange":1000000000}]'
  volatile.last_state.idmap: '[{"Isuid":true,"Isgid":false,"Hostid":1000000,"Nsid":0,"Maprange":1000000000},{"Isuid":false,"Isgid":true,"Hostid":1000000,"Nsid":0,"Maprange":1000000000}]'
  volatile.last_state.power: RUNNING
  volatile.lxbr31.host_name: veth7205e6b6
  volatile.lxbr31.hwaddr: 00:16:3e:04:8e:30
  volatile.lxbr31.name: eth0
devices:
  lxbr31:
    nictype: bridged
    parent: lxbr31
    type: nic
  root:
    path: /
    pool: svc-pool
    type: disk
  www-wetter:
    path: /www-wetter
    source: /root-docs/www-wetter
    type: disk
ephemeral: false
profiles:
- default
stateful: false
description: ""
 - [ ] Main daemon log (at /var/log/lxd/lxd.log or /var/snap/lxd/common/lxd/logs/lxd.log)
 - [ ] Output of the client with --debug
 - [ ] Output of the daemon with --debug (alternatively output of `lxc monitor` while reproducing the issue)

location: none
metadata:
  context: {}
  level: dbug
  message: 'Failure for task operation: 3ce222d5-8883-4a17-a380-72ad875d8740: Failed to start device "fuellstand-port": Can''t create device as devices path is mounted nodev'
timestamp: "2020-10-12T22:56:09.293230316+02:00"
type: logging

location: none
metadata:
  context: {}
  level: dbug
  message: 'Scheduler: container meteo changed: re-balancing'
timestamp: "2020-10-12T22:56:09.293255761+02:00"
type: logging

location: none
metadata:
  class: task
  created_at: "2020-10-12T22:56:09.285029763+02:00"
  description: Updating container
  err: 'Failed to start device "fuellstand-port": Can''t create device as devices path is mounted nodev'
  id: 3ce222d5-8883-4a17-a380-72ad875d8740
  location: none
  may_cancel: false
  metadata: null
  resources:
    containers:
    - /1.0/containers/meteo
  status: Failure
  status_code: 400
  updated_at: "2020-10-12T22:56:09.285029763+02:00"
timestamp: "2020-10-12T22:56:09.293689825+02:00"
type: operation

**Comment by stgraber** [comment]
You may need to reboot the system for this change to propagate into all mount namespaces.

**Comment by mhansi12** [comment]
I was able to "solve" the problem by creating a dedicated ZFS dataset for /var/snap and added the "devices=on" option.

After rebooting I checked the mountoptions of the relevant  mountpoints ( /var and /var/snap), /var was set to "devices=off" again and marked as "temporary". 

Currently I do not know the Ubuntu/ZFS boot process enough to understand who sets this option for /var

But anywhy thanks for your clarification.
